#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "cpu_filters.h"
#include <QFileDialog>
#include <QThread>
#include <QMessageBox>
#include <QImage>
#include <QPixmap>
#include <chrono>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    connect(ui->filterComboBox, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &MainWindow::on_filterComboBox_currentIndexChanged);
    connect(ui->gaussKernelSpin, QOverload<int>::of(&QSpinBox::valueChanged),
            this, &MainWindow::on_gaussianKernelSizeSpinBox_valueChanged);
    connect(ui->medianKernelSpin, QOverload<int>::of(&QSpinBox::valueChanged),
            this, &MainWindow::on_medianKernelSizeSpinBox_valueChanged);

}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::on_gaussianKernelSizeSpinBox_valueChanged(int value)
{
    if (value % 2 == 0) {
        ui->gaussKernelSpin->setValue(value + 1);
    }
}

void MainWindow::on_medianKernelSizeSpinBox_valueChanged(int value)
{
    if (value % 2 == 0) {
        ui->medianKernelSpin->setValue(value +1);
    }
}


void MainWindow::on_filterComboBox_currentIndexChanged(int index)
{
    // –ü—Ä–æ—Å—Ç–æ –ø–µ—Ä–µ–∫–ª—é—á–∞–µ–º —Å—Ç—Ä–∞–Ω–∏—Ü—É –≤ stackedWidget
    ui->parametersStack->setCurrentIndex(index);
}


void MainWindow::on_loadButton_clicked()
{
    QString filename = QFileDialog::getOpenFileName(this, "–û—Ç–∫—Ä—ã—Ç—å –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ –∏–ª–∏ –≤–∏–¥–µ–æ", "",
                                                    "–ò–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è –∏ –≤–∏–¥–µ–æ (*.png *.jpg *.bmp *.mp4 *.avi *.mov)");
    if (filename.isEmpty()) return;

    QString extension = QFileInfo(filename).suffix().toLower();

    if (extension == "png" || extension == "jpg" || extension == "bmp") {
        originalImage = cv::imread(filename.toStdString());
        if (originalImage.empty()) {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ.");
            return;
        }

        isVideo = false; // üëà —É–∫–∞–∑—ã–≤–∞–µ–º, —á—Ç–æ —ç—Ç–æ –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ

        cv::cvtColor(originalImage, originalImage, cv::COLOR_BGR2RGB);
        QImage qimg(originalImage.data, originalImage.cols, originalImage.rows, originalImage.step, QImage::Format_RGB888);
        ui->imageLabel->setPixmap(QPixmap::fromImage(qimg).scaled(ui->imageLabel->size(), Qt::KeepAspectRatio));
    }
    else if (extension == "mp4" || extension == "avi" || extension == "mov") {
        currentVideoPath = filename; // —Å–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ –ø—É—Ç–∏ –∫ –≤–∏–¥–µ–æ
        loadAndProcessVideo(filename);
        isVideo = true;
    }
    else {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ–ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ–º—ã–π —Ñ–æ—Ä–º–∞—Ç —Ñ–∞–π–ª–∞.");
    }
}

void MainWindow::loadAndProcessVideo(QString filePath)
{
    cv::VideoCapture cap(filePath.toStdString());
    if (!cap.isOpened()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –æ—Ç–∫—Ä—ã—Ç—å –≤–∏–¥–µ–æ.");
        return;
    }

    cv::Mat frame;
    while (cap.read(frame)) {
        if (frame.empty()) break;

        cv::Mat filteredFrame;
        originalImage = frame;  // —á—Ç–æ–±—ã —Ñ–∏–ª—å—Ç—Ä—ã –º–æ–≥–ª–∏ —Ä–∞–±–æ—Ç–∞—Ç—å —Å originalImage

        applySelectedFilter();  // —Ç–≤–æ—è —É–∂–µ –≥–æ—Ç–æ–≤–∞—è —Ñ—É–Ω–∫—Ü–∏—è
        filteredFrame = processedImage;

        if (filteredFrame.channels() == 1) {
            cv::cvtColor(filteredFrame, filteredFrame, cv::COLOR_GRAY2RGB);
        } else {
            cv::cvtColor(filteredFrame, filteredFrame, cv::COLOR_BGR2RGB);
        }

        QImage qimg(filteredFrame.data, filteredFrame.cols, filteredFrame.rows, static_cast<int>(filteredFrame.step), QImage::Format_RGB888);
        ui->imageLabel->setPixmap(QPixmap::fromImage(qimg).scaled(ui->imageLabel->size(), Qt::KeepAspectRatio));
        cv::waitKey(30); // –∑–∞–¥–µ—Ä–∂–∫–∞ –º–µ–∂–¥—É –∫–∞–¥—Ä–∞–º–∏ (–º–æ–∂–Ω–æ —Ä–µ–≥—É–ª–∏—Ä–æ–≤–∞—Ç—å)
        QCoreApplication::processEvents();  // –ø–æ–∑–≤–æ–ª—è–µ—Ç –æ–±–Ω–æ–≤–∏—Ç—å GUI
    }

    cap.release();
}



void MainWindow::on_saveButton_clicked()
{
    if (!isVideo) {
        if (processedImage.empty()) {
            QMessageBox::information(this, "–ù–µ—Ç –¥–∞–Ω–Ω—ã—Ö", "–°–Ω–∞—á–∞–ª–∞ –æ–±—Ä–∞–±–æ—Ç–∞–π—Ç–µ –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ.");
            return;
        }

        QString filename = QFileDialog::getSaveFileName(this, "–°–æ—Ö—Ä–∞–Ω–∏—Ç—å –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ", "", "PNG (*.png);;JPEG (*.jpg)");
        if (!filename.isEmpty()) {
            cv::Mat saveImg;
            cv::cvtColor(processedImage, saveImg, cv::COLOR_RGB2BGR);
            cv::imwrite(filename.toStdString(), saveImg);
        }
    } else {
        if (processedVideoFrames.empty()) {
            QMessageBox::information(this, "–ù–µ—Ç –¥–∞–Ω–Ω—ã—Ö", "–°–Ω–∞—á–∞–ª–∞ –ø—Ä–∏–º–µ–Ω–∏—Ç–µ —Ñ–∏–ª—å—Ç—Ä –∫ –≤–∏–¥–µ–æ.");
            return;
        }

        QString filename = QFileDialog::getSaveFileName(this, "–°–æ—Ö—Ä–∞–Ω–∏—Ç—å –≤–∏–¥–µ–æ", "", "–í–∏–¥–µ–æ (*.avi)");
        if (filename.isEmpty()) return;

        int fps = 30;
        cv::Size frameSize = processedVideoFrames[0].size();
        cv::VideoWriter writer(filename.toStdString(), cv::VideoWriter::fourcc('M','J','P','G'), fps, frameSize);

        for (const auto& frame : processedVideoFrames) {
            cv::Mat bgrFrame;
            cv::cvtColor(frame, bgrFrame, cv::COLOR_RGB2BGR);
            writer.write(bgrFrame);
        }

        writer.release();
    }
}


void MainWindow::on_applyButton_clicked()
{
    if (!isVideo) {
        // –°—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–∞—è –ª–æ–≥–∏–∫–∞ –¥–ª—è –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è
        if (originalImage.empty()) {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–°–Ω–∞—á–∞–ª–∞ –∑–∞–≥—Ä—É–∑–∏—Ç–µ –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ.");
            return;
        }

        auto start = std::chrono::high_resolution_clock::now();
        applySelectedFilter();
        auto end = std::chrono::high_resolution_clock::now();

        double duration = std::chrono::duration<double>(end - start).count();
        ui->timeLabel->setText(QString("–í—Ä–µ–º—è –æ–±—Ä–∞–±–æ—Ç–∫–∏: %1 —Å").arg(duration, 0, 'f', 6));

        cv::Mat rgb;

        QString selectedFilter = ui->filterComboBox->currentText();

        if (selectedFilter == "–û–ø–µ—Ä–∞—Ç–æ—Ä –°–æ–±–µ–ª—è" || selectedFilter == "–û–ø–µ—Ä–∞—Ç–æ—Ä –ö–∞–Ω–Ω–∏" || selectedFilter == "–ú–µ–¥–∏–∞–Ω–Ω—ã–π —Ñ–∏–ª—å—Ç—Ä") {
            cv::Mat normalized;
            cv::normalize(processedImage, normalized, 0, 255, cv::NORM_MINMAX);
            normalized.convertTo(normalized, CV_8U);
            cv::cvtColor(normalized, rgb, cv::COLOR_GRAY2RGB);
            QImage qimg(rgb.data, rgb.cols, rgb.rows, static_cast<int>(rgb.step), QImage::Format_RGB888);
            ui->imageLabel->setPixmap(QPixmap::fromImage(qimg).scaled(ui->imageLabel->size(), Qt::KeepAspectRatio, Qt::SmoothTransformation));
        } else {
            if (processedImage.channels() == 3) {
                cv::cvtColor(processedImage, rgb, cv::COLOR_BGR2RGB);
                QImage qimg(processedImage.data, processedImage.cols, processedImage.rows, processedImage.step, QImage::Format_RGB888);
                ui->imageLabel->setPixmap(QPixmap::fromImage(qimg).scaled(ui->imageLabel->size(), Qt::KeepAspectRatio));
            } else {
                QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ–ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ–º—ã–π —Ñ–æ—Ä–º–∞—Ç –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è.");
                return;
            }
        }
    } else {
        // –õ–æ–≥–∏–∫–∞ –¥–ª—è –≤–∏–¥–µ–æ
        processedVideoFrames.clear(); // –æ—á–∏—â–∞–µ–º —Å—Ç–∞—Ä—ã–µ –∫–∞–¥—Ä—ã

        cv::VideoCapture cap(currentVideoPath.toStdString());
        if (!cap.isOpened()) {
            QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –æ—Ç–∫—Ä—ã—Ç—å –≤–∏–¥–µ–æ.");
            return;
        }

        auto start = std::chrono::high_resolution_clock::now();

        cv::Mat frame;
        while (cap.read(frame)) {
            if (frame.empty()) break;

            originalImage = frame;
            applySelectedFilter();
            processedVideoFrames.push_back(processedImage.clone()); // —Å–æ—Ö—Ä–∞–Ω—è–µ–º –æ–±—Ä–∞–±–æ—Ç–∞–Ω–Ω—ã–π –∫–∞–¥—Ä
        }

        auto end = std::chrono::high_resolution_clock::now();
        double duration = std::chrono::duration<double>(end - start).count();
        ui->timeLabel->setText(QString("–í—Ä–µ–º—è –æ–±—Ä–∞–±–æ—Ç–∫–∏: %1 —Å").arg(duration, 0, 'f', 6));

        cap.release();

        // –í–æ—Å–ø—Ä–æ–∏–∑–≤–µ—Å—Ç–∏ –æ–±—Ä–∞–±–æ—Ç–∞–Ω–Ω–æ–µ –≤–∏–¥–µ–æ
        playProcessedVideo();
    }
}

void MainWindow::playProcessedVideo()
{
    for (const auto& frame : processedVideoFrames) {
        cv::Mat rgbFrame;
        if (frame.channels() == 1) {
            cv::cvtColor(frame, rgbFrame, cv::COLOR_GRAY2RGB);
        } else {
            cv::cvtColor(frame, rgbFrame, cv::COLOR_BGR2RGB);
        }

        QImage qimg(rgbFrame.data, rgbFrame.cols, rgbFrame.rows, static_cast<int>(rgbFrame.step), QImage::Format_RGB888);
        ui->imageLabel->setPixmap(QPixmap::fromImage(qimg).scaled(ui->imageLabel->size(), Qt::KeepAspectRatio));

        QCoreApplication::processEvents();
        QThread::msleep(30); // –∑–∞–¥–µ—Ä–∂–∫–∞ –¥–ª—è –∏–º–∏—Ç–∞—Ü–∏–∏ –≤–∏–¥–µ–æ
    }
}



void MainWindow::applySelectedFilter()
{
    QString filter = ui->filterComboBox->currentText();
    QString impl = ui->implementationComboBox->currentText();

    if (filter == "–†–∞–∑–º—ã—Ç–∏–µ –ø–æ –ì–∞—É—Å—Å—É") {
        int k = ui->gaussKernelSpin->value();
        double sigma = ui->gaussSigmaSpin->value();

        if (impl == "OpenCV")
            gaussianBlurCPU(originalImage, processedImage, k, sigma);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            gaussianBlurManual(originalImage, processedImage, k, sigma);
        else if (impl == "SIMD")
            gaussianBlurSIMD(originalImage, processedImage, k, sigma);
    }
    else if (filter == "–û–ø–µ—Ä–∞—Ç–æ—Ä –°–æ–±–µ–ª—è") {
        if (impl == "OpenCV")
            sobelFilterOpenCV(originalImage, processedImage);
        if (originalImage.empty()) {
            qDebug() << "Input image is empty!";
            return;
        }
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            sobelFilterManual(originalImage, processedImage);
        else if (impl == "SIMD")
            sobelFilterSIMD(originalImage, processedImage);
    }
    else if (filter == "–û–ø–µ—Ä–∞—Ç–æ—Ä –ö–∞–Ω–Ω–∏") {
        double low = ui->cannyLowSpin->value();
        double high = ui->cannyHighSpin->value();

        if (impl == "OpenCV")
            cannyEdgeDetectorOpenCV(originalImage, processedImage, low, high);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            cannyEdgeDetectorManual(originalImage, processedImage, low, high);
        else if (impl == "SIMD")
            cannyEdgeDetectorSIMD(originalImage, processedImage, low, high);
    }
    else if (filter == "–ú–µ–¥–∏–∞–Ω–Ω—ã–π —Ñ–∏–ª—å—Ç—Ä") {
        int kernelSize = ui->medianKernelSpin->value();

        if (impl == "OpenCV")
            medianFilterOpenCV(originalImage, processedImage, kernelSize);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            medianFilterManual(originalImage, processedImage, kernelSize);
        else if (impl == "SIMD")
            medianFilterSIMD(originalImage, processedImage, kernelSize);
    }
    else if (filter == "–ò–∑–º–µ–Ω–µ–Ω–∏–µ —è—Ä–∫–æ—Å—Ç–∏") {
        int beta = ui->brightBetaSpin->value();

        if (impl == "OpenCV")
            adjustBrightnessOpenCV(originalImage, processedImage, beta);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            adjustBrightnessManual(originalImage, processedImage, beta);
        else if (impl == "SIMD")
            adjustBrightnessSIMD(originalImage, processedImage, beta);
    }
    else if (filter == "–ò–∑–º–µ–Ω–µ–Ω–∏–µ –Ω–∞—Å—ã—â–µ–Ω–Ω–æ—Å—Ç–∏") {
        double contrastAlphaSpin = ui->contrastAlphaSpin->value();

        if (impl == "OpenCV")
            adjustSaturationOpenCV(originalImage, processedImage, contrastAlphaSpin);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            adjustSaturationManual(originalImage, processedImage, contrastAlphaSpin);
        else if (impl == "SIMD")
            adjustSaturationSIMD(originalImage, processedImage, contrastAlphaSpin);
    }
    else if (filter == "–ü–∏–∫—Å–µ–ª–∏–∑–∞—Ü–∏—è") {
        int pixelSize = ui->pixelSizeSpin->value();

        if (impl == "OpenCV")
            pixelateOpenCV(originalImage, processedImage, pixelSize);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            pixelateManual(originalImage, processedImage, pixelSize);
        else if (impl == "SIMD")
            pixelateSIMD(originalImage, processedImage, pixelSize);
    }
    else if (filter == "–ò–∑–º–µ–Ω–µ–Ω–∏–µ —Ä–µ–∑–∫–æ—Å—Ç–∏") {
        if (impl == "OpenCV")
            sharpenOpenCV(originalImage, processedImage);
        else if (impl == "–ë–µ–∑ –±–∏–±–ª–∏–æ—Ç–µ–∫")
            sharpenManual(originalImage, processedImage);
        else if (impl == "SIMD")
            sharpenSIMD(originalImage, processedImage);
    }
}
